src/models/polygon_utils.py

```diff
@@ -0,0 +1,143 @@
+"""Helper functions for overlap computations with polygons in shapely"""
+from numpy import ndarray
+from shapely.geometry.polygon import Polygon
+from geopandas.geoseries import GeometryArray
+
+# Note: All DE-9IM patterns below are streamlined to work well with polygons.
+#  They are not guaranteed to work on lower dimensional objects (points/lines)
+CORNER_ONLY_PATTERN = "FF*F0****"
+EDGE_ONLY_PATTERN = "FF*F1****"
+# Note: since we deal with polygons only, we can use a simplified overlap pattern:
+#  If polygons overlap with more than just their edge, they will automatically overlap
+#  with their interiors
+OVERLAP_PATTERN = "T********"
+
+
+def de9im_match(pattern: str, target_pattern: str) -> bool:
```

Comment:
I feel like there has to be a way to make this much more efficient, maybe through stuff like `numba` or Cython, but that's something to leave for later.

Reply:
Side note: when searching for what the relate function does I came to this: https://github.com/libgeos/geos/blob/186bbd32fbf07d8b5d419cdfd64c14e2270a418a/src/geom/IntersectionMatrix.cpp#L89 which I think are the functions which a lot of Shapely stuff calls.

Reply:
Ah nice, I didn't see the clibrary implementation - yes there's definitely a way to make it faster!

Overall, I haven't concentrated on this part much bc the string comparison is much faster than the overlap computations and rtree query so the overall speedup in the function would be only around 5%. 


But I agree that we should note this for later 👍 --> I'll convert it to an issue

Reply:
Ah ok, I wouldn't have guessed it was such a small part of the overall time just from the code. Probably not a priority then.